### DISCUSSION TITLE
Mobile wallet technology stack

### DISCUSSION THREAD

**Discussion by holzeis** [itchysats/10101#5] [General]
# Mobile wallet technology stack

Going for a mobile wallet we need to make a decision on the technology stack to be used. In general there are the following options.

## Cross platform frameworks based on typescript.

There are multiple framework available for developing a hybrid mobile app which runs on Android and iOS. In the blow comments I give a short summary about the most popular options. Lets separate our discussion on the different options - and eventually decide on one option.

## Native development 
Another option also mentioned here for completeness is to develop the mobile app completely native for Android or iOS. Performance and Security are definitely advantages of this option, but I don't think they are standing in comparison to the added effort and complexity.

## Additional resources

- [react.native VS Ionic](https://appinventiv.com/blog/react-native-vs-ionic/)
- [flutter VS react.native](https://appinventiv.com/blog/flutter-vs-react-native/)
- https://appinventiv.com/blog/cross-platform-app-frameworks/

**Comment by holzeis** [comment]
### [React Native](https://reactnative.dev/)
React Native lets you create truly native apps and doesn't compromise your users' experiences. It provides a core set of platform agnostic native components like `View`, `Text`, and `Image` that map directly to the platform’s native UI building blocks. 

> If you need plugins, strong community, glitch-free performance even with large data then React Native will provide that.

**Advantages**:
- Better performance than e.g. ionic as it is not running in a simple web view.
- Strong community support
- hot reload feature
- Written in Javascript / Typescript / JSX

**Disadvantages**:
- Lack of ui components, makes the UI development more complex for react.native
- Can only be tested on the mobile phone (or on the computer using emulated devices in android studio).

**Integrate Rust**
Yes, if you can trust the following medium post. https://medium.com/@marekkotewicz/building-a-mobile-app-in-rust-and-react-native-part-2-hello-world-60970a7d194a

**Reply by bonomat** [reply]
Advantage
+ it has the most developers which would make it easier to hire/outsource if we want to


> **Disadvantages**:
>     * Can only be tested on the mobile phone (or on the computer using emulated devices in android studio).
> 

I think you can build for desktop (windows + macOS)  as well: https://microsoft.github.io/react-native-windows/

**Reply by holzeis** [reply]
> > **Disadvantages**:
> > * Can only be tested on the mobile phone (or on the computer using emulated devices in android studio).
> 
> I think you can build for desktop (windows + macOS) as well: https://microsoft.github.io/react-native-windows/

It looks like it, but if you build for a mobile app - you have to run it at least in an emulated environment - unlike ionic, which is running in any browser.

**Comment by holzeis** [comment]
### [Flutter](https://flutter.dev/)
Flutter is an open source framework by Google for building beautiful, natively compiled, multi-platform applications from a single codebase.

> Flutter, backed by Google, delivers more of native experience than React Native and does not rely much on third party libraries for introducing new functionalities. This gives Flutter an upper hand over React Native

**Integrate Rust**
Yes, using the [flutter rust bridge](https://github.com/fzyzcjy/flutter_rust_bridge). see also https://dev.to/sunshine-chain/dart-meets-rust-a-match-made-in-heaven-9f5

**Advantages**:
- Does not need a bridge for communicating with native components.
- Is destined to replace react.native in the future.

**Disadvantages**:
- Written in dart.
- The least mature in this list.

**Reply by holzeis** [reply]
Flutter it is!

**Comment by holzeis** [comment]
### [Ionic](https://ionicframework.com/)
An open source mobile UI toolkit for building modern, high quality cross-platform mobile apps (run in a browser) from a single code base in (react).

> If you need less cost, easy to build with excellent designs and great user interface then Ionic framework will be the best option.

**Advantages**:
- Can be tested in any browser, as the app will also run in the browser on the phone.
- Is compatible with any web framework including react (but also angular and vue). So it should allow us to reuse existing components.
- A large set of ui components optimized for mobile (https://ionicframework.com/docs/components)

**Disadvantages**:
- Runs in a web view and thus is less efficient than other frameworks like react.native.

**Integrate Rust**
Less can be found on the internet regarding that aspect, but I guess cross compiling with neon-bindings should work - similar how we do it in our electron app. However, initial attempts on porting our app to mobile failed due to cross compiling issues.

**Comment by klochowicz** [comment]
After the latest academy session, where we looked at Flutter in depth and shared our experience with trying to get React Native to work (as well as going through other alternatives: Ionic, PWA, full native), we voted that Flutter is the way to go for us. 

I am archiving this discussion and proceeding with setting up the Flutter / Rust project using `flutter_rust_bridge`.
